### ISSUE TITLE
DT 4.4.2 crashing when creating GoPro GPR Raw thumbnails

### ISSUE THREAD

**Issue by bengtfalke** [darktable-org/darktable#15022]
### Describe the bug

I am using DT 4.4.2 installed from Flatpack on Linux Mint 21.2 on the hardware below. When I open a folder with lots of photos and scroll with the mouse wheel among the thumbnails in the Lighttable view… DT suddenly crashes? Is this a known problem?

### Steps to reproduce

1. Select Folder with lots of photos
2. Select one of the photos in the Lighttable view
3. Roll the mousewheel down in the list
4. Suddenlu after maybe 100 thumbnails DT crashes hard

### Logfile | Screenshot | Screencast

[DT crash 230807.txt](https://github.com/darktable-org/darktable/files/12279267/DT.crash.230807.txt)

### Where did you install darktable from?

flatpak

### darktable version

4.4.2

### What OS are you using?

Linux

### What is the version of your OS?

LInux Mint 21.2

### Describe your system?

CPU: AMD Ryzen 9 3900 (24) @ 3.100GH
GPU: NVIDIA GeForce RTX 2070 SUPER
Memory: 4773MiB / 32025MiB
SSD1: WD: WDS100T3X0C-00SJG0 size: 931.51 GiB
SSD2: Intel: SSDPEKNW020T8 size: 1.86 TiB
Linux Mint 21.2 with kernel 6.2 low latency

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

NVIDIA GeForce RTX 2070 SUPER

**Comment by gi-man** [comment]
As explained in pixls.us, your raw files seem to have a corruption. https://discuss.pixls.us/t/dt-4-4-2-crashing-when-scrolling-thumbnails/38761/17

**Comment by victoryforce** [comment]
@bengtfalke It would be great if you could find the single file that causes dt to crash when creating a thumbnail (if a problem with some image is really the cause) and share that file for bug tracking.

**Comment by bengtfalke** [comment]
This is solved now, thanks to Claes in Lund! It was the Flatpak version that was the cause of all these problems!

**Comment by victoryforce** [comment]
> It was the Flatpak version that was the cause of all these problems!

It's good that the native package doesn't seem to have this problem. But it is not very good that the build in which this problem occurs is possible.

I tried to check the current Windows build on the GoPro Hero 6 files from raw.pixls.us. I also observed the program crashes, but they were somewhat non-deterministic, not on all files, and much worse, unreproducible when run under the debugger. 🙁

> Do you want me to send you an example file?

If this is a single file on which flatpak dt is guaranteed to _always crash_, then yes, I would be interested to check (maybe, by the way, the Windows build is also vulnerable). It is better not to send this file to me personally, but to zip it and attach it to this issue.

**Comment by bengtfalke** [comment]
I will send two files tomorrow,,,, one CR2 file and one GPR... they are in the same folder here and when I did open that folder in the collection... DT crashed at once.

I uninstalled the Flatpak version and installed the version on GIT and all works perfectly with OpenCL running and so on.

Regards/Bengt

**Comment by bengtfalke** [comment]
Attached are one GPR file in ZIP format. When I open that folder DT would crash hard every tim
[GOPR0638.zip](https://github.com/darktable-org/darktable/files/12313567/GOPR0638.zip)
e.

**Comment by victoryforce** [comment]
Thank you for the sample. As I already wrote in the flatpak issue, I repeat it here, none of the native 4.4.2 builds have any problems with this file. But the same 4.4.2 in flatpak crashes on it guaranteed.

Crashing flatpak dt gives an assertion error in rawspeed, while this error is not present in other builds. But if it is the same release version of dt, then the rawspeed code should be identical. I have no explanation for this...

**Comment by hfiguiere** [comment]
>  I repeat it here, none of the native 4.4.2 builds have any problems with this file. But the same 4.4.2 in flatpak crashes on it guaranteed.

The crash (it's an abort()) is triggered by a c++ assertion. Flatpak builds with assertions enabled. (it's not a debug build)

```
/usr/include/c++/13.2.0/optional:477: constexpr _Tp& std::_Optional_base_impl<_Tp, _Dp>::_M_get() [with _Tp = rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> >; _Dp = std::_Optional_base<rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> >, false, false>]: Assertion 'this->_M_is_engaged()' failed.
```

So if it reaches it, it's a bug. It's not Flatpak's fault, and closing it without change is the wrong resolution. You just hide the problem by ignoring assertions.

The assert is triggered by dereferencing `this->codeDecoder` that is of type `std::optional<PrefixCodeDecoder>`.

From the C++ reference:

<https://en.cppreference.com/w/cpp/utility/optional/operator*>

> The behavior is undefined if *this does not contain a value. 

So it's a bug. Why doesn't it contain a value ? No idea, thats either a bug or expected. Should the code check that it contains a value before dereferencing? Definitely. Why does it work without assertion enabled? It's coincidental, it's just by chance, or it's initialized to something. But it's **undefined behaviour**. That's the bug.

**Comment by hfiguiere** [comment]
The stack trace for posterity. This is the 4.6.0 Flatpak.

```
#0  0x00007ffff76a3e14 in __pthread_kill_implementation () at /usr/lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff7651dce in raise () at /usr/lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff763983f in abort () at /usr/lib/x86_64-linux-gnu/libc.so.6
#3  0x00007ffff38ececd in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff7c848df in std::_Optional_base_impl<rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> >, std::_Optional_base<rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> >, false, false> >::_M_get() (this=0x7fffcffe9898) at /usr/include/c++/13.2.0/optional:475
#5  std::_Optional_base_impl<rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> >, std::_Optional_base<rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> >, false, false> >::_M_get()
    (this=0x7fffcffe9898) at /usr/include/c++/13.2.0/optional:475
#6  std::optional<rawspeed::PrefixCodeLUTDecoder<rawspeed::VC5CodeTag, rawspeed::PrefixCodeVectorDecoder<rawspeed::VC5CodeTag> > >::operator*() &
    (this=0x7fffcffe9898) at /usr/include/c++/13.2.0/optional:974
#7  rawspeed::VC5Decompressor::parseLargeCodeblock(rawspeed::ByteStream) (this=this@entry=0x7fffcffe9850, bs=...)
    at /run/build/darktable/src/external/rawspeed/src/librawspeed/decompressors/VC5Decompressor.cpp:793
#8  0x00007ffff7c84dd6 in rawspeed::VC5Decompressor::parseVC5() (this=this@entry=0x7fffcffe9850)
    at /run/build/darktable/src/external/rawspeed/src/librawspeed/io/ByteStream.h:118
#9  0x00007ffff7c8562d in rawspeed::VC5Decompressor::VC5Decompressor(rawspeed::ByteStream, rawspeed::RawImage const&)
    (this=this@entry=0x7fffcffe9850, bs=..., img=...) at /run/build/darktable/src/external/rawspeed/src/librawspeed/decompressors/VC5Decompressor.cpp:428
#10 0x00007ffff7c63f5a in rawspeed::AbstractDngDecompressor::decompressThread<9>() const (this=<optimized out>)
    at /run/build/darktable/src/external/rawspeed/src/librawspeed/decompressors/AbstractDngDecompressor.cpp:151
#11 0x00007ffff7e8da07 in GOMP_parallel () at /usr/lib/x86_64-linux-gnu/libgomp.so.1
#12 0x00007ffff7c63306 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Alloc_hider::_Alloc_hider(char*, std::allocator<char>&&) (__a=<optimized out>, __dat=<optimized out>, this=<optimized out>) at /usr/include/c++/13.2.0/bits/basic_string.h:193
#13 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string() (this=<optimized out>)
    at /usr/include/c++/13.2.0/bits/basic_string.h:514
#14 rawspeed::AbstractDngDecompressor::decompress() const (this=this@entry=0x7fffcffe9de0)
    at /run/build/darktable/src/external/rawspeed/src/librawspeed/decompressors/AbstractDngDecompressor.cpp:223
#15 0x00007ffff7c27d4c in rawspeed::DngDecoder::decodeData(rawspeed::TiffIFD const*, unsigned int) const
    (this=this@entry=0x7fffc809ca40, raw=raw@entry=0x7fffc80cf110, sample_format=sample_format@entry=1)
    at /run/build/darktable/src/external/rawspeed/src/librawspeed/decoders/DngDecoder.cpp:419
#16 0x00007ffff7c282ed in rawspeed::DngDecoder::decodeRawInternal() (this=0x7fffc809ca40)
    at /run/build/darktable/src/external/rawspeed/src/librawspeed/decoders/DngDecoder.cpp:503
#17 0x00007ffff7c47a5a in rawspeed::RawDecoder::decodeRaw() (this=0x7fffc809ca40)
```

**Comment by kmilos** [comment]
@LebedevRI this is for rawspeed then, not dt.

**Comment by kmilos** [comment]
Thanks. I presume you won't be merging all those changes to your stable branch any time soon (nor dt backporting anything to 4.6.x), so I guess it's for 4.8 then... @TurboGit

**Comment by LebedevRI** [comment]
Eh, there hasn't been any really breaking changes there,
so i think it would be a good idea to update the stable branch
before proceeding into e.g. the black level changes.

**Comment by LebedevRI** [comment]
Fixed in dt 4.6.1
